lib/services/unified_resource.go

```diff
@@ -128,6 +128,26 @@ func (c *UnifiedResourceCache) put(ctx context.Context, resource resource) error
 	c.mu.Lock()
 	defer c.mu.Unlock()
 	key := resourceKey(resource)
+	oldResource, exists := c.resources[key]
+	if exists && resource.GetKind() == types.KindNode {
+		oldHostname := oldResource.(types.Server).GetHostname()
+		newHostname := resource.(types.Server).GetHostname()
+		if oldHostname != newHostname {
```

Comment:
This is making an assumption about how the resource sort keys are constructed which is currently abstracted away by `makeResourceSortKey()`. As an alternative, you could use the resource sort key as the determinator of whether to update the btrees or not, which would also fix this for all resource types if it ever becomes an issue. I'm thinking something like this (untested):

```
        key := resourceKey(resource)
        sortKey := makeResourceSortKey(resource)

        if oldResource, exists := c.resources[key]; exists {
                // If the resource has changed in such a way that the sort keys
                // for the nameTree or typeTree change, remove the old entries
                // from those trees before adding a new one. This can happen for
                // nodes at least where the hostname can change.
                oldSortKey := makeResourceSortKey(oldResource)
                if oldSortKey.byName != sortKey.byName {
                        if _, ok := c.nameTree.Delete(&item{Key: sortKey.byName}); !ok {
                                return trace.NotFound("key %q is not found in unified cache name sort tree", string(sortKey.byName))
                }
                if oldSortKey.byType != sortKey.byType {
                        if _, ok := c.typeTree.Delete(&item{Key: sortKey.byType}); !ok {
                                return trace.NotFound("key %q is not found in unified cache type sort tree", string(sortKey.byType))
                        }
                }
        }

        c.resources[key] = resource
        c.nameTree.ReplaceOrInsert(&item{Key: sortKey.byName, Value: key})
        c.typeTree.ReplaceOrInsert(&item{Key: sortKey.byType, Value: key})
 ```

I didn't transfer your comment since I didn't understand it completely (only because I'm not familiar with what is going on here in detail, not for any reasons of the message itself), so I'd also ensure that the important information of that comment is kept.

Reply:
I like this in theory a lot more. Let me test it out and I'll update.